rfc/20250317-ephemeral-resources.md

```diff
@@ -0,0 +1,646 @@
+* [Prevent write-only for sets](https://github.com/hashicorp/terraform-plugin-framework/pull/1095)
+* [Nullifying write-only attributes moved to an earlier stage](https://github.com/hashicorp/terraform-plugin-framework/pull/1097)
+
+On the OpenTofu side the following needs to be tackled:
+* Update [Attribute](https://github.com/opentofu/opentofu/blob/ff4c84055065fa2d83d318155b72aef6434d99e4/internal/configs/configschema/schema.go#L44) to add a field for the WriteOnly flag.
+* Update the validation of the provider generated plan in such a way to allow nil values for the fields that are actually having a value defined in the configuration. This is necessary because the plugin framework is setting nil any values that are marked as write-only.
+  * Test this in-depth for all the block types except sets of any kind (Investigate and understand why sets are not allowed by the plugin framework).
+    * Add a new validation on the provider schema to check against, set nested attributes and set nested blocks with writeOnly=true. Tested this with a version of terraform-plugin-framework that allowed writeOnly on sets and there is an error returned. (set attributes are allowed based on my tests)
+    In order to understand this better, maybe we should allow this for the moment and test OpenTofu with the [plugin-framework version](https://github.com/hashicorp/terraform-plugin-framework/commit/0724df105602e6b6676e201b7c0c5e1d187df990) that allows sets to be write-only=true.
+
+> [!NOTE]
+>
+> Write-only attributes will be presented in the OpenTofu's UI as `(write-only attribute)` instead of the actual value.
+
+### Variables
+
+For enabling ephemeral variables, these are the basic steps that need to be taken:
+* Update config to support the `ephemeral` attribute.
+* Mark the variables with a new mark ensure that the marks are propagated correctly.
+* Based on the marks, ensure that the variable cannot be used in other contexts than the ephemeral ones (see the User Documentation section for more details on where this is allowed).
+* Check the state of [#1998](https://github.com/opentofu/opentofu/pull/1998). If that is merged, in the changes where variables from plan are verified against the configuration ones, we also need to add a validation on the ephemerality of variables. If the variable is marked as ephemeral, then the plan value is allowed (expected) to be missing. 
+* Ensure that when the prompt is shown for an ephemeral variable there is an indication of that:
+  ```hcl
+  var.password (ephemeral)
+     Enter a value:
+  ````
+* If a module is having an ephemeral variable declared, that variable can get values from any source, even another non-ephemeral variable.
+
+We should use boolean marks, as no additional information is required to be carried. When introducing the marks for these, extra care should be taken in *all* the places marks are handled and ensure that the existing implementation around marks is not affected.
```

Comment:
do we also need to introduce some checks/tests to ensure that functions such as `sensitive()` do not accidentally strip the ephemeral mark too?

Reply:
3fe9247caedeef0099138111dece7a18065cda2b